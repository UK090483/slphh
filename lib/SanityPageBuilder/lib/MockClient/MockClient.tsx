import type { SanityClient } from "@sanity/client/sanityClient";
import { parse, evaluate } from "groq-js";

type MockSanityClient = {
  mockReturnValue?: any;
  database?: any;
};

export const mockClient = (props?: MockSanityClient) => {
  const { mockReturnValue, database } = props || {
    mockReturnValue: null,
    database: null,
  };

  return {
    fetch: (query: string) => {
      if (mockReturnValue) {
        return Promise.resolve(mockReturnValue);
      }

      return fetchMock(database, query);
    },
  } as unknown as SanityClient;
};

export const mockGetClient = (props: MockSanityClient) => {
  return () => mockGetClient(props);
};

const fetchMock = async (dataset: any, query: string) => {
  let tree = parse(query);
  let value = await evaluate(tree, { dataset });
  return await value.get();
};
