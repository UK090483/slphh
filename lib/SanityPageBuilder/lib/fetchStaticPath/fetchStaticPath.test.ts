import fetchStaticPaths from "./fetchStaticPath";
import { mockClient } from "../MockClient/MockClient";

const locales = {
  de: { title: "Deutsch", isDefault: true, flag: "ðŸ‡©ðŸ‡ª" },
  en: { title: "Englisch", flag: "ðŸ‡ºðŸ‡¸" },
  da: { title: "DÃ¤nisch", flag: "ðŸ‡©ðŸ‡°" },
};

const database = [
  ...["testPageHome", "testPage1", "widthSubRoute2", "CAPITAL-TestPage3"].map(
    (i) => ({
      _id: i,
      _type: "page",
      pageType: i === "widthSubRoute2" ? { _ref: "testPageType" } : null,
      slug: { current: i },
      slug_en: { current: i + "en" },
      slug_da: { current: i + "da" },
    })
  ),

  {
    _id: "noSlug",
    _type: "page",
    slug_en: { current: "noSlugEn" },
    slug_da: { current: "noSlugEn" },
  },
  {
    _id: "noTransSlug",
    _type: "page",
    slug: { current: "noTransSlug" },
  },

  { _id: "testPageType", _type: "pageType", slug: { current: "testPageType" } },
  { _id: "siteConfig", indexPage: { _ref: "testPageHome" } },
];

const mockFetchStaticPaths = ({
  fetchReturn,
  database,
  locales,
  fallback,
}: {
  fetchReturn?: any;
  database?: any;
  locales?: any;
  fallback?: boolean | "blocking";
}) => {
  return fetchStaticPaths({
    doc: "page",
    client: mockClient({
      ...(fetchReturn && { mockReturnValue: fetchReturn }),
      ...(database && { database }),
    }),
    locales: locales,
    fallback,
  });
};

describe("fetchStaticPath", () => {
  it("should trow error if fetch result is not [] ", async () => {
    await expect(mockFetchStaticPaths({ fetchReturn: null })).rejects.toThrow();
    await expect(mockFetchStaticPaths({ fetchReturn: "st" })).rejects.toThrow();
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] } })
    ).resolves.toStrictEqual({ fallback: "blocking", paths: [] });
  });

  it("should handle fallback ", async () => {
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] } })
    ).resolves.toStrictEqual({ fallback: "blocking", paths: [] });
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] }, fallback: false })
    ).resolves.toStrictEqual({ fallback: false, paths: [] });
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] }, fallback: true })
    ).resolves.toStrictEqual({ fallback: true, paths: [] });
  });

  it("should return right data no locales ", async () => {
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] } })
    ).resolves.toStrictEqual({ fallback: "blocking", paths: [] });
    await expect(mockFetchStaticPaths({ database })).resolves.toMatchSnapshot();
  });

  it("should return right data ", async () => {
    await expect(
      mockFetchStaticPaths({ fetchReturn: { slugs: [] }, locales })
    ).resolves.toStrictEqual({ fallback: "blocking", paths: [] });

    await expect(
      mockFetchStaticPaths({ database, locales })
    ).resolves.toMatchSnapshot();
  });
});
