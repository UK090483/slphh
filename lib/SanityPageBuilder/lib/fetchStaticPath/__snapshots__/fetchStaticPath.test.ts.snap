// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchStaticPath should return right data  1`] = `
Object {
  "fallback": "blocking",
  "paths": Array [
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "testpagehome",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "testpagehomeen",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "testpagehomeda",
        ],
      },
    },
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "testpage1",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "testpage1en",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "testpage1da",
        ],
      },
    },
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "testpagetype",
          "widthsubroute2",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "testpagetype",
          "widthsubroute2en",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "testpagetype",
          "widthsubroute2da",
        ],
      },
    },
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "capital-testpage3",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "capital-testpage3en",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "capital-testpage3da",
        ],
      },
    },
    Object {
      "locale": "de",
      "params": Object {
        "slug": Array [
          "notransslug",
        ],
      },
    },
    Object {
      "locale": "en",
      "params": Object {
        "slug": Array [
          "notransslug",
        ],
      },
    },
    Object {
      "locale": "da",
      "params": Object {
        "slug": Array [
          "notransslug",
        ],
      },
    },
  ],
}
`;

exports[`fetchStaticPath should return right data no locales  1`] = `
Object {
  "fallback": "blocking",
  "paths": Array [
    Object {
      "params": Object {
        "slug": Array [
          "testpagehome",
        ],
      },
    },
    Object {
      "params": Object {
        "slug": Array [
          "testpage1",
        ],
      },
    },
    Object {
      "params": Object {
        "slug": Array [
          "testpagetype",
          "widthsubroute2",
        ],
      },
    },
    Object {
      "params": Object {
        "slug": Array [
          "capital-testpage3",
        ],
      },
    },
    Object {
      "params": Object {
        "slug": Array [
          "notransslug",
        ],
      },
    },
  ],
}
`;
